feature: add remove flag @pulumi/pulumi/automation

Moon1706 opened the pull request:
<!--- 
Thanks so much for your contribution! If this is your first time contributing, please ensure that you have read the [CONTRIBUTING](https://github.com/pulumi/pulumi/blob/master/CONTRIBUTING.md) documentation.
-->

# Description
This PR is just added functionality which is already existed in Pulumi 3.43.0 to automation NPM package which use in Pulumi Github Action.

<!--- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. -->

Fixes: [issue](https://github.com/pulumi/pulumi/issues/11078)
PR: [main functionality](https://github.com/pulumi/pulumi/pull/10943)

## Checklist

<!--- Please provide details if the checkbox below is to be left unchecked. -->
- [x] _I have added tests that prove my fix is effective or that my feature works_. Your tests aren't working now (path: `sdk/nodejs`, command: `make test_auto`)
<!--- 
User-facing changes require a CHANGELOG entry.
-->
- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change
<!--
If the change(s) in this PR is a modification of an existing call to the Pulumi Service,
then the service should honor older versions of the CLI where this change would not exist.
You must then bump the API version in /pkg/backend/httpstate/client/api.go, as well as add
it to the service.
-->
- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version
  <!-- @Pulumi employees: If yes, you must submit corresponding changes in the service repo. -->

kpitzen commented:
/run-acceptance-tests

Moon1706 commented:
@kpitzen 
Hey! I'm wondering why in your message I see a valid pipeline, but in this PR workflow shows a problem with comment and gather(

RobbieMcKinstry commented:
> Hey! I'm wondering why in your message I see a valid pipeline, but in this PR workflow shows a problem with comment and gather
@Moon1706 I think the build is failing because it's from a fork. It looks like a bug in the CI job; we recently switched to an awesome new CI build system and haven't worked out all the kinks. I wouldn't worry about it.

After discussing with @kpitzen , we want to keep the Automation API libraries in parity with each other. Any interest in adding the code to support Python, Go, and C# too? NBD if not, we're grateful for your contribution either way.

RobbieMcKinstry commented:
BTW once we think this PR is ready for review, we can submit it for merge with bors even if the `gather` check fails, because Bors will assert everything passes independent of this check.

Moon1706 commented:
@RobbieMcKinstry 
I've added. Check please

RobbieMcKinstry reviewed (approved):
This looks good to me. @kpitzen did you say you wanted four unit tests (one per lang) before merging? Or does this look good to you?

RobbieMcKinstry commented:
Oh, it also needs a CHANGELOG entry. From the top level directory, you can run `make changelog` and follow the prompt to generate a new entry @Moon1706

kpitzen commented:
> This looks good to me. @kpitzen did you say you wanted four unit tests (one per lang) before merging? Or does this look good to you?

Hi @Moon1706 !  This all looks very good to me!  Per @RobbieMcKinstry 's comment above, unit tests would be great - I don't know if I'd go so far as to request one per language, but we do have some automation test suites for node: https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/tests/automation/localWorkspace.spec.ts and python: https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/test/automation/test_local_workspace.py  If you could add a test in either (or both) of these languages, I think that would _likely_ cover our case here.  Given Automation API is currently just calls out to the engine, that should be sufficient in this case.  Thank you again for your contribution!  Very much looking forward to getting this in :)

Moon1706 commented:
> Oh, it also needs a CHANGELOG entry. From the top level directory, you can run `make changelog` and follow the prompt to generate a new entry @Moon1706

Hi! I've already done this) Now I'll see the testing section

Moon1706 commented:
> > This looks good to me. @kpitzen did you say you wanted four unit tests (one per lang) before merging? Or does this look good to you?
> 
> Hi @Moon1706 ! This all looks very good to me! Per @RobbieMcKinstry 's comment above, unit tests would be great - I don't know if I'd go so far as to request one per language, but we do have some automation test suites for node: https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/tests/automation/localWorkspace.spec.ts and python: https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/test/automation/test_local_workspace.py If you could add a test in either (or both) of these languages, I think that would _likely_ cover our case here. Given Automation API is currently just calls out to the engine, that should be sufficient in this case. Thank you again for your contribution! Very much looking forward to getting this in :)

Sure. No problem) But I faced with strange issue. Maybe it's because I've executed tests on the local machine, but:
``` js
    it(`runs through the stack lifecycle with an inline program`, asyncTest(async () => {
        const program = async () => {
            const config = new Config();
            return {
                exp_static: "foo",
                exp_cfg: config.get("bar"),
                exp_secret: config.getSecret("buzz"),
            };
        };
        const projectName = "inline_node";
        const stackName = fullyQualifiedStackName(getTestOrg(), projectName, `int_test${getTestSuffix()}`);
        const stack = await LocalWorkspace.createStack({ stackName, projectName, program });

        const stackConfig: ConfigMap = {
            "bar": { value: "abc" },
            "buzz": { value: "secret", secret: true },
        };
        await stack.setAllConfig(stackConfig);

        // pulumi up
        const upRes = await stack.up({ userAgent });
        assert.strictEqual(Object.keys(upRes.outputs).length, 3);
        assert.strictEqual(upRes.outputs["exp_static"].value, "foo");
        assert.strictEqual(upRes.outputs["exp_static"].secret, false);
        assert.strictEqual(upRes.outputs["exp_cfg"].value, "abc");
        assert.strictEqual(upRes.outputs["exp_cfg"].secret, false);
        assert.strictEqual(upRes.outputs["exp_secret"].value, "secret");
        assert.strictEqual(upRes.outputs["exp_secret"].secret, true);
        assert.strictEqual(upRes.summary.kind, "update");
        assert.strictEqual(upRes.summary.result, "succeeded");

        // pulumi preview
        const preRes = await stack.preview({ userAgent });
        assert.strictEqual(preRes.changeSummary.same, 1);

        // pulumi refresh
        const refRes = await stack.refresh({ userAgent });
        assert.strictEqual(refRes.summary.kind, "refresh");
        assert.strictEqual(refRes.summary.result, "succeeded");

        // pulumi destroy
        const destroyRes = await stack.destroy({ userAgent });
        assert.strictEqual(destroyRes.summary.kind, "destroy");
        assert.strictEqual(destroyRes.summary.result, "succeeded");

        // pulumi destroy with remove flag
        // const ws = stack.workspace;
        const remove = true;
        await stack.destroy({ userAgent, remove });
        // const allStacksWs = await ws.listStacks();
        // const findStacks = allStacksWs.find((stackCheck: StackSummary) => stackCheck.name === stackName);
        // assert(!findStacks);
```
I've added the last 6 lines and seen the following issue:
```sh
  1) LocalWorkspace
       runs through the stack lifecycle with an inline program:
     StackNotFoundError: code: -2
 stdout:
 stderr: Command failed with exit code 255: pulumi stack history --json --show-secrets --page-size 1 --page 1 --stack pulumi-test-inline_node-int_test252422 --non-interactive
error: no stack named 'pulumi-test-inline_node-int_test252422' found
 err?: Error: Command failed with exit code 255: pulumi stack history --json --show-secrets --page-size 1 --page 1 --stack pulumi-test-inline_node-int_test252422 --non-interactive
error: no stack named 'pulumi-test-inline_node-int_test252422' found

      at Object.createCommandError (automation/errors.ts:23:53)
      at Object.<anonymous> (bin/automation/cmd.js:17:1727)
      at Generator.throw (<anonymous>)
      at rejected (bin/automation/cmd.js:15:987)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)



error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
make: *** [Makefile:72: test_auto] Error 1
```
Please, could you help me with this, because I don't pretty good understand why I've got this error?

kpitzen commented:
Good morning (in my timezone!) @Moon1706 !  From what I see in the test code you've provided, the lines you've added come _after_ these lines:
```typescript
        // pulumi destroy
        const destroyRes = await stack.destroy({ userAgent });
        assert.strictEqual(destroyRes.summary.kind, "destroy");
        assert.strictEqual(destroyRes.summary.result, "succeeded");
```
This tells me that the test has _already_ destroyed the stack that your test code would then also try to destroy, so the stack likely doesn't exist anymore.  To test this, I would add a new test similar to the one above, but which does a single destroy using the new `remove` flag.  Thanks again for adding this!

Moon1706 commented:
@kpitzen 
Hi! Unfortunately, I can't agree with you, because:
1. Previously we've deleted stack using the line `await stack.workspace.removeStack(stack.name);` which I've rewritten.
2. I've changed code to 
```js
    it(`runs through the stack lifecycle with an inline program`, asyncTest(async () => {
        const program = async () => {
            const config = new Config();
            return {
                exp_static: "foo",
                exp_cfg: config.get("bar"),
                exp_secret: config.getSecret("buzz"),
            };
        };
        const projectName = "inline_node";
        const stackName = fullyQualifiedStackName(getTestOrg(), projectName, `int_test${getTestSuffix()}`);
        const stack = await LocalWorkspace.createStack({ stackName, projectName, program });

        const stackConfig: ConfigMap = {
            "bar": { value: "abc" },
            "buzz": { value: "secret", secret: true },
        };
        await stack.setAllConfig(stackConfig);

        // pulumi up
        const upRes = await stack.up({ userAgent });
        assert.strictEqual(Object.keys(upRes.outputs).length, 3);
        assert.strictEqual(upRes.outputs["exp_static"].value, "foo");
        assert.strictEqual(upRes.outputs["exp_static"].secret, false);
        assert.strictEqual(upRes.outputs["exp_cfg"].value, "abc");
        assert.strictEqual(upRes.outputs["exp_cfg"].secret, false);
        assert.strictEqual(upRes.outputs["exp_secret"].value, "secret");
        assert.strictEqual(upRes.outputs["exp_secret"].secret, true);
        assert.strictEqual(upRes.summary.kind, "update");
        assert.strictEqual(upRes.summary.result, "succeeded");

        // pulumi preview
        const preRes = await stack.preview({ userAgent });
        assert.strictEqual(preRes.changeSummary.same, 1);

        // pulumi refresh
        const refRes = await stack.refresh({ userAgent });
        assert.strictEqual(refRes.summary.kind, "refresh");
        assert.strictEqual(refRes.summary.result, "succeeded");

        // pulumi destroy
        const remove = true;
        const destroyRes = await stack.destroy({ userAgent, remove });
        assert.strictEqual(destroyRes.summary.kind, "destroy");
        assert.strictEqual(destroyRes.summary.result, "succeeded");

        // pulumi destroy with remove flag
        // const ws = stack.workspace;
        // const remove = true;
        // await stack.destroy({ userAgent, remove });
        // const allStacksWs = await ws.listStacks();
        // const findStacks = allStacksWs.find((stackCheck: StackSummary) => stackCheck.name === stackName);
        // assert(!findStacks);
    }));
```
and got the same issue.
